File: backend/src/models/Clinic.js
import mongoose from 'mongoose';
import { TIME_RE, DATE_RE } from '../utils/clinicTime.js';

const timeValidator = { validator: (v) => TIME_RE.test(v), message: 'time must be HH:mm (24h)' };

/**
 * One recurring window in the weekly schedule — e.g. Monday 10:00–14:00. A day
 * can have several (a morning and an evening sitting), so this is a flat list
 * keyed by dayOfWeek rather than one entry per day.
 */
const weeklyHoursSchema = new mongoose.Schema(
  {
    dayOfWeek: { type: Number, min: 0, max: 6, required: true }, // 0 = Sunday
    start: { type: String, required: true, validate: timeValidator },
    end: { type: String, required: true, validate: timeValidator },
  },
  { _id: false },
);

const windowSchema = new mongoose.Schema(
  {
    start: { type: String, required: true, validate: timeValidator },
    end: { type: String, required: true, validate: timeValidator },
  },
  { _id: false },
);

/**
 * A date-specific exception to the weekly pattern: a holiday closure
 * (`isClosed`) or special one-off hours (`windows`). Overrides win over the
 * weekly schedule for that date.
 */
const overrideSchema = new mongoose.Schema(
  {
    date: { type: String, required: true, match: DATE_RE }, // 'YYYY-MM-DD'
    isClosed: { type: Boolean, default: false },
    windows: { type: [windowSchema], default: [] },
    note: { type: String, maxlength: 200 },
  },
  { _id: false },
);

/**
 * A physical location where the doctor holds consultations, together with the
 * schedule of when they are available there. Doctor and staff manage these; the
 * slot engine ([services/scheduling.js]) turns the schedule into bookable
 * times, and patients book against it.
 */
const clinicSchema = new mongoose.Schema(
  {
    name: { type: String, required: true, trim: true, maxlength: 160 },
    addressLine: { type: String, trim: true, maxlength: 400 },
    city: { type: String, trim: true, maxlength: 120 },
    phone: { type: String, trim: true, maxlength: 40 },
    // A clinic usually publishes more than one line, and a patient who cannot
    // get through on the first should not have to hunt for the second.
    altPhone: { type: String, trim: true, maxlength: 40 },
    // Optional map link (Google Maps, etc.) so a patient can find the place.
    mapUrl: { type: String, trim: true, maxlength: 600 },

    // ---- Brand ------------------------------------------------------------
    //
    // The clinic's identity as the patient meets it: on the chat header, on the
    // prescription letterhead, in an appointment confirmation. This lived in
    // two env vars, which meant renaming the clinic took a redeploy — and made
    // a second clinic impossible, since one process can only hold one value.
    //
    // MedPin is the product; this is the tenant. The app's own name stays on
    // the icon, the splash and the About screen, and everything a patient reads
    // carries the clinic instead.

    /// The line under the name — "Diabetes Obesity & Metabolic Clinic".
    tagline: { type: String, trim: true, maxlength: 160 },

    /// The doctor's name as it should be printed, which is not always the name
    /// on their account ("Dr. Amit Kumar Dey" vs how they sign).
    doctorDisplayName: { type: String, trim: true, maxlength: 160 },

    /// Two logo assets, not one, and never an inverted copy of the other.
    ///
    /// Inverting artwork to fit a background destroys the brand colour — this
    /// clinic's teal comes out orange. So the two variants are stored as the
    /// designer drew them: `logoLight` is the one that reads on a light
    /// surface, `logoDark` the one for a dark surface.
    ///
    /// The app is light-only today, so `logoLight` is the one it draws. A
    /// clinic that only ever supplies dark-background artwork still renders —
    /// see `logoNeedsDarkChip`.
    logoLightAssetId: { type: mongoose.Schema.Types.ObjectId, ref: 'MediaAsset' },
    logoDarkAssetId: { type: mongoose.Schema.Types.ObjectId, ref: 'MediaAsset' },

    /// Set when the only artwork supplied is drawn for a dark background.
    ///
    /// Measured on upload from the mean luminance of the non-transparent
    /// pixels, not guessed. When true the app paints the logo on a dark rounded
    /// chip rather than inverting it: the brand colours survive and the mark
    /// stays legible on a white screen.
    logoNeedsDarkChip: { type: Boolean, default: false },

    /// Printed under the signature on a prescription.
    registrationNo: { type: String, trim: true, maxlength: 60 },

    // The doctor whose availability this schedule represents. Single-doctor
    // today; the ref keeps a multi-doctor build open.
    doctor: { type: mongoose.Schema.Types.ObjectId, ref: 'User', index: true },

    slotMinutes: { type: Number, default: 15, min: 5, max: 120 },
    weeklyHours: { type: [weeklyHoursSchema], default: [] },
    overrides: { type: [overrideSchema], default: [] },

    // Soft-disable rather than delete, so past appointments keep their clinic.
    isActive: { type: Boolean, default: true, index: true },
    sortIndex: { type: Number, default: 0 },
  },
  { timestamps: true },
);

clinicSchema.methods.toPublic = function toPublic() {
  return {
    id: this._id,
    name: this.name,
    addressLine: this.addressLine ?? null,
    city: this.city ?? null,
    phone: this.phone ?? null,
    altPhone: this.altPhone ?? null,
    mapUrl: this.mapUrl ?? null,
    tagline: this.tagline ?? null,
    doctorDisplayName: this.doctorDisplayName ?? null,
    registrationNo: this.registrationNo ?? null,
    logoLightUrl: this.logoLightAssetId ? `/api/v1/uploads/${this.logoLightAssetId}/raw` : null,
    logoDarkUrl: this.logoDarkAssetId ? `/api/v1/uploads/${this.logoDarkAssetId}/raw` : null,
    logoNeedsDarkChip: Boolean(this.logoNeedsDarkChip),
    slotMinutes: this.slotMinutes,
    weeklyHours: (this.weeklyHours ?? [])
      .map((w) => ({ dayOfWeek: w.dayOfWeek, start: w.start, end: w.end }))
      .sort((a, b) => a.dayOfWeek - b.dayOfWeek || a.start.localeCompare(b.start)),
    overrides: (this.overrides ?? []).map((o) => ({
      date: o.date,
      isClosed: o.isClosed,
      windows: (o.windows ?? []).map((w) => ({ start: w.start, end: w.end })),
      note: o.note ?? null,
    })),
    isActive: this.isActive,
    sortIndex: this.sortIndex,
    createdAt: this.createdAt,
  };
};

/** Lightweight shape for embedding on an appointment. */
clinicSchema.methods.toBrief = function toBrief() {
  return {
    id: this._id,
    name: this.name,
    addressLine: this.addressLine ?? null,
    city: this.city ?? null,
    phone: this.phone ?? null,
  };
};

export const Clinic = mongoose.model('Clinic', clinicSchema);

File: mobile/lib/features/appointments/domain/clinic.dart
/// A recurring weekly availability window, e.g. Monday 10:00–14:00.
class WeeklyHour {
  const WeeklyHour({
    required this.dayOfWeek,
    required this.start,
    required this.end,
  });

  /// 0 = Sunday … 6 = Saturday (matches the backend and Dart's DateTime.weekday
  /// modulo 7).
  final int dayOfWeek;
  final String start; // 'HH:mm'
  final String end; // 'HH:mm'

  factory WeeklyHour.fromJson(Map<String, dynamic> j) => WeeklyHour(
    dayOfWeek: (j['dayOfWeek'] as num?)?.toInt() ?? 0,
    start: j['start']?.toString() ?? '00:00',
    end: j['end']?.toString() ?? '00:00',
  );

  Map<String, dynamic> toJson() => {
    'dayOfWeek': dayOfWeek,
    'start': start,
    'end': end,
  };
}

/// A date-specific exception to the weekly pattern — a holiday closure or
/// special one-off hours.
class ClinicOverride {
  const ClinicOverride({
    required this.date,
    required this.isClosed,
    this.windows = const [],
    this.note,
  });

  final String date; // 'YYYY-MM-DD'
  final bool isClosed;
  final List<({String start, String end})> windows;
  final String? note;

  factory ClinicOverride.fromJson(Map<String, dynamic> j) => ClinicOverride(
    date: j['date']?.toString() ?? '',
    isClosed: j['isClosed'] == true,
    windows:
        (j['windows'] as List?)
            ?.whereType<Map<String, dynamic>>()
            .map(
              (w) => (
                start: w['start']?.toString() ?? '',
                end: w['end']?.toString() ?? '',
              ),
            )
            .toList() ??
        const [],
    note: j['note']?.toString(),
  );

  Map<String, dynamic> toJson() => {
    'date': date,
    'isClosed': isClosed,
    'windows': [
      for (final w in windows) {'start': w.start, 'end': w.end},
    ],
    if (note != null) 'note': note,
  };
}

/// A physical location where the doctor consults, with its schedule.
class Clinic {
  const Clinic({
    required this.id,
    required this.name,
    this.addressLine,
    this.city,
    this.phone,
    this.altPhone,
    this.mapUrl,
    this.tagline,
    this.doctorDisplayName,
    this.registrationNo,
    this.logoLightUrl,
    this.logoDarkUrl,
    this.logoNeedsDarkChip = false,
    this.slotMinutes = 15,
    this.weeklyHours = const [],
    this.overrides = const [],
    this.isActive = true,
    this.sortIndex = 0,
    this.managedByYou,
  });

  final String id;
  final String name;
  final String? addressLine;
  final String? city;
  final String? phone;

  /// A second line patients can ring. A clinic usually publishes more than one,
  /// and somebody who cannot get through on the first should not have to hunt.
  final String? altPhone;
  final String? mapUrl;

  // ---- Brand ---------------------------------------------------------------
  //
  // What a patient meets: on the chat header, on the prescription letterhead,
  // in an appointment confirmation. MedPin is the product; this is the clinic.

  /// The line under the name — "Diabetes Obesity & Metabolic Clinic".
  final String? tagline;

  /// The doctor's name as it should be printed, which is not always the name on
  /// their account.
  final String? doctorDisplayName;

  /// Printed under the signature on a prescription.
  final String? registrationNo;

  /// Two logos, never one inverted into the other.
  ///
  /// Inversion is a per-channel complement, so this clinic's teal comes back a
  /// muddy orange — and colour is the part of a logo that carries the brand.
  /// The app is light-only, so [logoLightUrl] is what it draws.
  final String? logoLightUrl;
  final String? logoDarkUrl;

  /// The only artwork supplied was drawn for a dark background.
  ///
  /// Measured on upload from the mean luminance of the non-transparent pixels.
  /// When true the app paints the mark on a dark chip rather than inverting it:
  /// brand colours survive and it stays legible on a white card.
  final bool logoNeedsDarkChip;
  final int slotMinutes;
  final List<WeeklyHour> weeklyHours;
  final List<ClinicOverride> overrides;
  final bool isActive;
  final int sortIndex;

  /// Whether the reader runs this location: may book, confirm, move and edit
  /// here. The server says so for staff and leaves it out for a patient, so
  /// null is "not a question for this reader", never "no".
  final bool? managedByYou;

  /// Open, and somewhere this reader may put an appointment.
  bool get takesWorkFromYou => isActive && managedByYou != false;

  /// Every number the clinic publishes, in the order it publishes them.
  List<String> get phones => [
    if (phone != null && phone!.isNotEmpty) phone!,
    if (altPhone != null && altPhone!.isNotEmpty) altPhone!,
  ];

  /// "DD-24, Salt Lake City · Kolkata" — a one-line location summary.
  String get locationLine {
    final parts = [
      if (addressLine != null && addressLine!.isNotEmpty) addressLine,
      if (city != null && city!.isNotEmpty) city,
    ];
    return parts.join(' · ');
  }

  factory Clinic.fromJson(Map<String, dynamic> j) => Clinic(
    id: j['id']?.toString() ?? '',
    name: j['name']?.toString() ?? '',
    addressLine: j['addressLine']?.toString(),
    city: j['city']?.toString(),
    phone: j['phone']?.toString(),
    altPhone: j['altPhone']?.toString(),
    mapUrl: j['mapUrl']?.toString(),
    tagline: j['tagline']?.toString(),
    doctorDisplayName: j['doctorDisplayName']?.toString(),
    registrationNo: j['registrationNo']?.toString(),
    logoLightUrl: j['logoLightUrl']?.toString(),
    logoDarkUrl: j['logoDarkUrl']?.toString(),
    logoNeedsDarkChip: j['logoNeedsDarkChip'] == true,
    slotMinutes: (j['slotMinutes'] as num?)?.toInt() ?? 15,
    weeklyHours:
        (j['weeklyHours'] as List?)
            ?.whereType<Map<String, dynamic>>()
            .map(WeeklyHour.fromJson)
            .toList() ??
        const [],
    overrides:
        (j['overrides'] as List?)
            ?.whereType<Map<String, dynamic>>()
            .map(ClinicOverride.fromJson)
            .toList() ??
        const [],
    isActive: j['isActive'] != false,
    sortIndex: (j['sortIndex'] as num?)?.toInt() ?? 0,
    managedByYou: j['managedByYou'] is bool ? j['managedByYou'] as bool : null,
  );
}

/// How many places an appointment could be put, which decides what to ask.
///
/// Location is optional. With none there is nothing to choose and no published
/// hours — the time is the doctor's to give. With one there is nothing to
/// choose either, so no picker is drawn. Only two or more is a question.
enum LocationChoice { none, one, several }

/// The locations this reader may put an appointment at, and what that means
/// for the screen. Counted among the ones they run, so a receptionist who runs
/// one branch of two is not asked which.
({List<Clinic> open, LocationChoice choice}) locationChoice(
  List<Clinic> all,
) {
  final open = all.where((c) => c.takesWorkFromYou).toList();
  return (
    open: open,
    choice: switch (open.length) {
      0 => LocationChoice.none,
      1 => LocationChoice.one,
      _ => LocationChoice.several,
    },
  );
}

/// One bookable slot returned by `GET /clinics/:id/slots`.
class Slot {
  const Slot({required this.time, required this.iso, required this.available});

  final String time; // 'HH:mm' clinic-local
  final String iso; // absolute ISO instant to send back when booking
  final bool available;

  factory Slot.fromJson(Map<String, dynamic> j) => Slot(
    time: j['time']?.toString() ?? '',
    iso: j['iso']?.toString() ?? '',
    available: j['available'] == true,
  );
}

/// The slot listing for a clinic on one date.
class SlotDay {
  const SlotDay({
    required this.clinicId,
    required this.date,
    required this.slotMinutes,
    required this.slots,
    this.isActive = true,
  });

  final String clinicId;
  final String date;
  final int slotMinutes;
  final List<Slot> slots;

  /// False for a closed location, whose empty list means closed rather than
  /// fully booked. Older servers do not send it, and read as open.
  final bool isActive;

  bool get hasAvailability => slots.any((s) => s.available);

  factory SlotDay.fromJson(Map<String, dynamic> j) => SlotDay(
    clinicId: j['clinicId']?.toString() ?? '',
    date: j['date']?.toString() ?? '',
    isActive: j['isActive'] != false,
    slotMinutes: (j['slotMinutes'] as num?)?.toInt() ?? 15,
    slots:
        (j['slots'] as List?)
            ?.whereType<Map<String, dynamic>>()
            .map(Slot.fromJson)
            .toList() ??
        const [],
  );
}
